docs/parser.md

```diff
@@ -0,0 +1,247 @@
+  "errors": {
+    "required/0": "This field is required and I'll tell you about it at the document level.",
+    "shallowlyRequired/required/0": "This deep field is required."
+  }
+}
+```
+
+# Requirements
+
+To use this component, you will need to instantiate it and make it aware of your schemas.  These will be dereferenced
+and cached as they are added.  You can then use the `evolveError` method to resolve the custom error definition for a
+field using the schema key and the field's JSON pointer.  See the validator implementation for an example of its use.
+
+This component uses [json-schema-ref-parser](https://github.com/BigstickCarpet/json-schema-ref-parser).  We configure
+it using `options.parserOptions` (empty by default).  Read the library's documentation for details of [what options
+are available](https://github.com/BigstickCarpet/json-schema-ref-parser/blob/master/docs/options.md).
+
+Note that this library expects to be able to resolve external references relative to the first argument passed to its
+`dereference` function, and it expects to load and cache the references itself as needed.  Since we are receiving keys
+from the `validator`, we need to some how convert these to URLs.  We do this using `fluid.stringTemplate` to combine
+`options.schemaPath` with the individual key, which means:
+
+1.  If `options.schemaPath` is a directory location, a full path will be produced.
+2.  If `options.schemaPath` is a URL, a full URL will be produced.
+3.  If `options.schemaPath` is empty, paths will be resolved relative to the working directory.
+
+# Components
+
+## `gpii.schema.parser`
+
+The base component is designed to work both on the client and server side.
+
+## `gpii.schema.parser.server`
+
+A server-side version of the component which can handle package-relative `schemaPath` entries like `%gpii-package-name/path/within/package`.
+
+# Functions
+
+## gpii.schema.parser.dereferenceSchema(that, schemaPath, schemaKey)
+
+ * `that`: The parser component itself.
+ * `schemaPath {String}`:  A path or URI to the directory or base URL where the schema can be found.
+ * `schemaKey {String}`: The filename of the schema relative to `schemaPath`.
+ * Returns: A `promise` that will be satisfied when the parser finishes its work.
+
+ Dereference a single schema.  As the process is asynchronous, this function returns a promise.  Typically accessed
+ using the parser component's `dereferenceSchema` invoker and the `schemaKey` attribute.
+
+## gpii.schema.parser.evolveError(that, schemaKey, error)
+
+* `that`: The parser component itself.
+* `schemaKey {String}`: The `key` of the schema (generally the filename, i.e. `base.json`).
+* `error {object}`: The original error returned by AJV.
+* Returns: `Object` The "evolved" error with an updated message based on the `errors` definitions (if there are any).
+
+This function used by the validator to transform the full set of raw validation results using `fluid.transform`.  It
```

Comment:
There was a little bit on the cloned return value further down, which I moved up and expanded.